File: api/disponibilidad.py
"""
Endpoint: /api/disponibilidad
Obtención de disponibilidad para las próximas 2 semanas
"""
from http.server import BaseHTTPRequestHandler
import json
import logging
from datetime import datetime, timedelta
from lib.citas import SistemaCitas

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        """Obtiene disponibilidad de las próximas 2 semanas"""
        try:
            sistema_citas = SistemaCitas()
            disponibilidad = {}
            hoy = datetime.now().date()

            # Obtener disponibilidad para los próximos 14 días
            for i in range(14):
                fecha = hoy + timedelta(days=i)
                fecha_str = fecha.strftime('%Y-%m-%d')

                horarios_info = sistema_citas.obtener_horarios_disponibles(fecha_str)

                # Crear lista de horarios ocupados
                todos_horarios = ['09:00', '10:00', '11:00', '12:00', '13:00', '14:00', '15:00', '16:00', '17:00', '18:00']
                horarios_disponibles_set = {h['hora'] for h in horarios_info.get('horarios_disponibles', [])}
                horarios_ocupados = [h for h in todos_horarios if h not in horarios_disponibles_set]

                disponibilidad[fecha_str] = {
                    'fecha': fecha_str,
                    'dia_semana': horarios_info.get('dia_semana', ''),
                    'horarios_ocupados': horarios_ocupados,
                    'horarios_disponibles': list(horarios_disponibles_set)
                }

            # Responder
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({
                'desde': hoy.strftime('%Y-%m-%d'),
                'hasta': (hoy + timedelta(days=13)).strftime('%Y-%m-%d'),
                'disponibilidad': disponibilidad
            }, ensure_ascii=False).encode('utf-8'))

        except Exception as e:
            logger.error(f"Error en /api/disponibilidad: {e}")
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({
                'error': f'Error obteniendo disponibilidad: {str(e)}'
            }).encode())

    def do_OPTIONS(self):
        """Manejo de preflight CORS"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

File: api/disponibilidad.js
/**
 * Vercel Serverless Function: /api/disponibilidad
 * Returns availability for the next 14 days
 * Replaces the Python version (disponibilidad.py)
 */

const HORARIO_INICIO = 9;
const HORARIO_FIN = 18;
const MAX_CITAS_POR_HORA = 1;
const DIAS_SEMANA = ['domingo', 'lunes', 'martes', 'miércoles', 'jueves', 'viernes', 'sábado'];

function generarTodosHorarios() {
  const horarios = [];
  for (let h = HORARIO_INICIO; h < HORARIO_FIN; h++) {
    horarios.push(`${String(h).padStart(2, '0')}:00`);
  }
  return horarios;
}

function formatDate(date) {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, '0');
  const d = String(date.getDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
}

/**
 * Get all booked hours from Google Calendar for a date range
 */
async function getCalendarEventsRange(startDate, endDate) {
  try {
    if (!process.env.GOOGLE_CREDENTIALS_JSON) return {};

    const { google } = await import('googleapis');
    const credentials = JSON.parse(process.env.GOOGLE_CREDENTIALS_JSON);

    const auth = new google.auth.GoogleAuth({
      credentials,
      scopes: ['https://www.googleapis.com/auth/calendar.readonly'],
    });

    const calendar = google.calendar({ version: 'v3', auth });
    const calendarId = process.env.GOOGLE_CALENDAR_ID || 'primary';
    const timezone = process.env.GOOGLE_CALENDAR_TIMEZONE || 'America/Mexico_City';

    const response = await calendar.events.list({
      calendarId,
      timeMin: new Date(`${startDate}T00:00:00`).toISOString(),
      timeMax: new Date(`${endDate}T23:59:59`).toISOString(),
      singleEvents: true,
      orderBy: 'startTime',
      timeZone: timezone,
    });

    const eventsByDate = {};
    for (const event of (response.data.items || [])) {
      const start = event.start?.dateTime;
      if (start) {
        const date = new Date(start);
        const dateStr = formatDate(date);
        const hora = `${String(date.getHours()).padStart(2, '0')}:00`;
        if (!eventsByDate[dateStr]) eventsByDate[dateStr] = [];
        eventsByDate[dateStr].push(hora);
      }
    }
    return eventsByDate;
  } catch (error) {
    console.error('Error getting calendar events:', error);
    return {};
  }
}

/**
 * Get all booked appointments from Google Sheets
 */
async function getSheetsAppointmentsAll() {
  try {
    if (!process.env.GOOGLE_CREDENTIALS_JSON || !process.env.GOOGLE_SHEET_ID) return {};

    const { google } = await import('googleapis');
    const credentials = JSON.parse(process.env.GOOGLE_CREDENTIALS_JSON);

    const auth = new google.auth.GoogleAuth({
      credentials,
      scopes: ['https://www.googleapis.com/auth/spreadsheets.readonly'],
    });

    const sheets = google.sheets({ version: 'v4', auth });
    const response = await sheets.spreadsheets.values.get({
      spreadsheetId: process.env.GOOGLE_SHEET_ID,
      range: 'Citas!A:H',
    });

    const rows = response.data.values || [];
    const appointmentsByDate = {};

    for (let i = 1; i < rows.length; i++) {
      const row = rows[i];
      if (row.length >= 8) {
        const fechaCita = (row[5] || '').trim();
        const horaCita = (row[6] || '').trim();
        const estado = (row[7] || '').trim().toLowerCase();

        if (fechaCita && estado === 'confirmada' && horaCita) {
          const parts = horaCita.replace(/\s*(AM|PM)\s*/gi, '').split(':');
          if (parts.length >= 1) {
            const h = parseInt(parts[0], 10);
            if (!isNaN(h)) {
              const normalizedHora = `${String(h).padStart(2, '0')}:00`;
              if (!appointmentsByDate[fechaCita]) appointmentsByDate[fechaCita] = [];
              appointmentsByDate[fechaCita].push(normalizedHora);
            }
          }
        }
      }
    }

    return appointmentsByDate;
  } catch (error) {
    console.error('Error getting sheets appointments:', error);
    return {};
  }
}

export default async function handler(req, res) {
  res.setHeader('Content-Type', 'application/json');
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  if (req.method !== 'GET') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const hoy = new Date();
    const todosHorarios = generarTodosHorarios();

    // Calculate date range
    const startDate = formatDate(hoy);
    const endDateObj = new Date(hoy);
    endDateObj.setDate(endDateObj.getDate() + 13);
    const endDate = formatDate(endDateObj);

    // Get all occupied hours from both sources in parallel
    const [calendarEvents, sheetsAppointments] = await Promise.all([
      getCalendarEventsRange(startDate, endDate),
      getSheetsAppointmentsAll()
    ]);

    // Build availability for each of the next 14 days
    const disponibilidad = {};

    for (let i = 0; i < 14; i++) {
      const fecha = new Date(hoy);
      fecha.setDate(fecha.getDate() + i);
      const fechaStr = formatDate(fecha);

      // Merge occupied hours from both sources
      const horasOcupadas = [
        ...(calendarEvents[fechaStr] || []),
        ...(sheetsAppointments[fechaStr] || [])
      ];

      // Count per hour
      const citasPorHorario = {};
      for (const hora of horasOcupadas) {
        citasPorHorario[hora] = (citasPorHorario[hora] || 0) + 1;
      }

      // Determine available and occupied
      const horariosOcupados = [];
      const horariosDisponibles = [];

      for (const horario of todosHorarios) {
        const count = citasPorHorario[horario] || 0;
        if (count >= MAX_CITAS_POR_HORA) {
          horariosOcupados.push(horario);
        } else {
          horariosDisponibles.push(horario);
        }
      }

      // If today, filter past hours from available
      if (i === 0) {
        const currentHour = hoy.getHours();
        const filteredDisponibles = horariosDisponibles.filter(h => {
          const hour = parseInt(h.split(':')[0], 10);
          return hour > currentHour;
        });

        disponibilidad[fechaStr] = {
          fecha: fechaStr,
          dia_semana: DIAS_SEMANA[fecha.getDay()],
          horarios_ocupados: horariosOcupados,
          horarios_disponibles: filteredDisponibles
        };
      } else {
        disponibilidad[fechaStr] = {
          fecha: fechaStr,
          dia_semana: DIAS_SEMANA[fecha.getDay()],
          horarios_ocupados: horariosOcupados,
          horarios_disponibles: horariosDisponibles
        };
      }
    }

    return res.status(200).json({
      desde: startDate,
      hasta: endDate,
      disponibilidad
    });

  } catch (error) {
    console.error('Error getting disponibilidad:', error);
    return res.status(500).json({
      error: `Error obteniendo disponibilidad: ${error.message}`
    });
  }
}
